src/Sylius/Bundle/ShopBundle/Grid/ProductGrid.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ShopBundle\Grid;
+
+use Sylius\Bundle\GridBundle\Builder\Field\DateTimeField;
+use Sylius\Bundle\GridBundle\Builder\Field\Field;
+use Sylius\Bundle\GridBundle\Builder\Field\StringField;
+use Sylius\Bundle\GridBundle\Builder\Filter\StringFilter;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
+use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Core\Model\TaxonInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Webmozart\Assert\Assert;
+
+final class ProductGrid extends AbstractGrid implements ResourceAwareGridInterface
+{
+    /**
+     * @param TaxonRepositoryInterface<TaxonInterface> $taxonRepository
+     */
+    public function __construct(
+        private string $resourceClass,
+        private ChannelContextInterface $channelContext,
+        private TaxonRepositoryInterface $taxonRepository,
+        private LocaleContextInterface $localeContext,
+        private RequestStack $requestStack,
+        private bool  $includeAllDescendants,
+    ) {
+    }
+
+    public static function getName(): string
+    {
+        return 'sylius_shop_product';
+    }
+
+    public function buildGrid(GridBuilderInterface $gridBuilder): void
+    {
+        $request = $this->requestStack->getMainRequest();
+        Assert::notNull($request, 'The '.self::class. ' does not work in CLI mode');
+
+        $localeCode = $this->localeContext->getLocaleCode();
+
+        // @see Sylius\Bundle\ResourceBundle\ExpressionLanguage\NotNullExpressionFunctionProvider
+        $taxon = $this->taxonRepository->findOneBySlug($request->attributes->get('slug'), $localeCode);
+        if ($taxon === null)  {
+            throw new NotFoundHttpException('Requested page is invalid');
+        }
+
+        $gridBuilder
+            ->setRepositoryMethod('createShopListQueryBuilder', [
+                $this->channelContext->getChannel(),
+                $taxon,
+                $localeCode,
+                $request->query->all('sorting'),
```

Comment:
Out of memory, you will need to specify fallback and possibly type cast, otherwise there will be cases, when it will fail.

Which it will be better to check for route parts before using it in repository

Reply:
This will only fail if the sorting is not an array, but then we would run into a type error in the repository method as well. If the sorting parameter isn't there it will default to an empty array. Which I think is a sensible default value.

Reply:
Got it, yes fallback to empty array is good default option.